block-buffer/src/lib.rs

```diff
@@ -1,48 +1,324 @@
+#[derive(Copy, Clone, Debug, Default)]
+pub struct Lazy {}
+
+impl BufferKind for Eager {}
+impl BufferKind for Lazy {}
+
+/// Eager block buffer.
+pub type EagerBuffer<B> = BlockBuffer<B, Eager>;
+/// Lazy block buffer.
+pub type LazyBuffer<B> = BlockBuffer<B, Lazy>;
+
+/// Buffer for block processing of data.
+#[derive(Debug)]
+pub struct BlockBuffer<BlockSize, Kind>
+where
+    BlockSize: ArrayLength<u8> + IsLess<U256>,
+    Le<BlockSize, U256>: NonZero,
+    Kind: BufferKind,
+{
+    buffer: Block<BlockSize>,
+    pos: u8,
+    _pd: PhantomData<Kind>,
+}
+
+impl<BlockSize, Kind> Default for BlockBuffer<BlockSize, Kind>
+where
+    BlockSize: ArrayLength<u8> + IsLess<U256>,
+    Le<BlockSize, U256>: NonZero,
+    Kind: BufferKind,
+{
+    fn default() -> Self {
+        Self {
+            buffer: Default::default(),
+            pos: 0,
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<BlockSize, Kind> Clone for BlockBuffer<BlockSize, Kind>
+where
+    BlockSize: ArrayLength<u8> + IsLess<U256>,
+    Le<BlockSize, U256>: NonZero,
+    Kind: BufferKind,
+{
+    fn clone(&self) -> Self {
+        Self {
+            buffer: self.buffer.clone(),
+            pos: self.pos,
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<BlockSize, Kind> BlockBuffer<BlockSize, Kind>
+where
+    BlockSize: ArrayLength<u8> + IsLess<U256>,
+    Le<BlockSize, U256>: NonZero,
+    Kind: BufferKind,
+{
+    /// Create new buffer from slice.
+    ///
+    /// # Panics
+    /// If slice length is not valid for used buffer kind.
+    #[inline(always)]
+    pub fn new(buf: &[u8]) -> Self {
+        let pos = buf.len();
+        assert!(Kind::invariant(pos, BlockSize::USIZE));
+        let mut buffer = Block::<BlockSize>::default();
+        buffer[..pos].copy_from_slice(buf);
+        Self {
+            buffer,
+            pos: pos as u8,
+            _pd: PhantomData,
+        }
+    }
+
     /// Digest data in `input` in blocks of size `BlockSize` using
     /// the `compress` function, which accepts slice of blocks.
-    fn digest_blocks(&mut self, input: &[u8], compress: impl FnMut(&[Block<BlockSize>]));
+    #[inline]
+    pub fn digest_blocks(
+        &mut self,
+        mut input: &[u8],
+        mut compress: impl FnMut(&[Block<BlockSize>]),
+    ) {
+        let pos = self.get_pos();
+        // using `self.remaining()` for some reason
+        // prevents panic elimination
+        let rem = self.size() - pos;
+        let n = input.len();
+        // Note that checking condition `pos + n < BlockSize` is
+        // equivalent to checking `n < rem`, where `rem` is equal
+        // to `BlockSize - pos`. Using the latter allows us to work
+        // around compiler accounting for possible overflow of
+        // `pos + n` which results in it inserting unreachable
+        // panic branches. Using `unreachable_unchecked` in `get_pos`
+        // we convince compiler that `BlockSize - pos` never underflows.
```

Comment:
I believe you mean overflow here, not underflow? If BlockSize - pos < 0, that's still overflow. Underflow only applies to floating point arithmetic.

Reply:
https://cwe.mitre.org/data/definitions/191.html

> CWE-191: Integer Underflow (Wrap or Wraparound)
>
> The product subtracts one value from another, such that the result is less than the minimum allowable integer value, which produces a value that is not equal to the correct result.

Reply:
I guess it is saying specifically "integer underflow" and not "arithmetic underflow", but that's a bit of a confusing usage imo.
https://en.wikipedia.org/wiki/Arithmetic_underflow
> Storing values that are too low in an [integer](https://en.wikipedia.org/wiki/Integer_(computer_science)) variable (e.g., attempting to store −1 in an [unsigned](https://en.wikipedia.org/wiki/Signedness) integer) is properly referred to as [integer overflow](https://en.wikipedia.org/wiki/Integer_overflow), or more broadly, integer wraparound. The term underflow normally refers to floating point numbers only, which is a separate issue.

Reply:
We're debugging some a crash around here (actually in `get_pos()`) that only happens only on win x86 in release config. It started only recently so it may be related to rustc version and it's a bit odd because it appears to crash on https://github.com/RustCrypto/utils/pull/671/files#diff-dab38e6f4d05085a1ab6b7b064fb097fa39f1806d8b03dfbb80472aaa21851e0R171
I'll open a ticket for it if we can repro in a pure rust test

Reply:
@bridiver 
If you do not trigger the debug assert on line 173 when you execute your program compiled in debug mode, then it's likely either a miscompilation (i.e. a compiler bug) or you have a memory corruption bug somewhere which damages block buffer memory.

Either way, reproduction will be appreciated. It may be worth to try inspect generated assembly in [Godbolt](https://rust.godbolt.org) (you can simply copy-paste `block-buffer` code into it with a bug reproduction code).

Reply:
> @bridiver If you do not trigger the debug assert on line 173 when you execute your program compiled in debug mode, then it's likely either a miscompilation (i.e. a compiler bug) or you have a memory corruption bug somewhere which damages block buffer memory.

There are no crashes in debug and this can be reliably reproduced across multiple builds by our tests. The tests passed in many previous releases and there are no changes to this particular code since the last successful run, but there are other changes like rustc version and possibly related changes to the build.

Reply:
Sorry, actually I take that back. I don't think we have tried a build that would trigger the debug asserts so we'll look at that as well.

Reply:
so we do see this with the debug assertions enabled
```
thread '<unnamed>' panicked at .\../../brave/third_party/rust/block_buffer/v0_10/crate/src/lib.rs:198:13:
assertion failed: false
```

Reply:
I still don't have a pure rust STR, but I'll open an issue

Reply:
btw - the panic message when we replicated this in rust was `attempt to subtract with overflow` :)